File: observer/python/myobserver.py
from abc import ABC, abstractmethod

class Subject(ABC):
    observers = []
    @abstractmethod
    def attach(self, observer):
        pass
    
    @abstractmethod
    def detach(self, observer):
        pass
        
    def notify(self, message):
        for observer in self.observers:
            observer.update(message)
            
class Observer():
    def update(self, message):
        print("Observer received message : " + message)
        
class Chef(Subject):
    def attach(self, observer):
        self.observers.append(observer)

    def detach(self, observer):
        self.observers.remove(observer)

    def prepareDish(self, dishname):
        self.notify("Preparation of dish " + dishname)
        
class Waiter(Observer):
    def __init__(self, chef):
        self.chef = chef
        self.chef.attach(self)
        
    def stopObserving(self):
        self.chef.detach(self)
        
    def update(self, message):
        print("Observer Waiter received message " + message)
        
def main():
    chef = Chef()
    waiter1 = Waiter(chef)
    
    chef.prepareDish(dishname = "TiraMisu")
    
    waiter1.stopObserving()
    
    chef.prepareDish(dishname = "Pasta")

if __name__ == "__main__":
    main()

File: observer/swift/myobserver.swift
import Foundation

// Observer Pattern Protocols

protocol Observer: AnyObject {
    var id: UUID { get }
    func update(message: String)
}

protocol Subject {
    var observers: Set<AnyObserverWrapper> { get set }
    func register(observer: Observer)
    func remove(observer: Observer)
    func notifyObservers(message: String)
}

// Wrapper to allow Set to store class-based Observers via Hashable
struct AnyObserverWrapper: Hashable {
    weak var observer: Observer?
    let id: UUID

    init(_ observer: Observer) {
        self.observer = observer
        self.id = observer.id
    }

    func hash(into hasher: inout Hasher) { hasher.combine(id) }
    static func == (lhs: AnyObserverWrapper, rhs: AnyObserverWrapper) -> Bool { lhs.id == rhs.id }
}

// Concrete Subject

class Chef: Subject {
    var observers = Set<AnyObserverWrapper>()
    private var lastOrder: String = ""

    func register(observer: Observer) {
        observers.insert(AnyObserverWrapper(observer))
    }

    func remove(observer: Observer) {
        observers.remove(AnyObserverWrapper(observer))
    }

    func notifyObservers(message: String) {
        observers.forEach { $0.observer?.update(message: message) }
    }

    func prepareOrder(_ dish: String) {
        print("Chef: Preparing \(dish)...")
        self.lastOrder = dish
        notifyObservers(message: "Order for \(dish) is ready!")
    }
}

// Concrete Observer

class Waiter: Observer {
    let id = UUID()
    let name: String
    private weak var chef: Chef?

    init(name: String, chef: Chef) {
        self.name = name
        self.chef = chef
        // Automatically register with the chef upon initialization
        chef.register(observer: self)
    }

    func update(message: String) {
        print("Waiter \(name) received notification: \(message)")
        serveOrder()
    }

    private func serveOrder() {
        print("Waiter \(name) is serving the dish to the table.")
    }
}

// Main Execution

let headChef = Chef()

let waiter1 = Waiter(name: "Alice", chef: headChef)
let waiter2 = Waiter(name: "Bob", chef: headChef)

print("--- Restaurant Service Started ---")
headChef.prepareOrder("Pasta Carbonara")

print("\n--- Bob leaves for break ---")
headChef.remove(observer: waiter2)

headChef.prepareOrder("Margherita Pizza")
